<script src='./fx.js'></script>

## 지연된 함수열을 병렬적으로 평가하기 - C.reduce, C.take

<script>
  /*
    자바스크립트 실행환경에서 (브라우저, node등) 보통 비동기 I/O로 동작한다.
    비동기 I/O라는 것은 싱글 스레드를 기반으로 I/O를 동기적으로 처리하기보다 비동기적으로 처리하여 하나의 스레드에서도 CPU를 효율적으로 사용한다.
    자바스크립트가 싱글스레드이다보니 병렬적인 작업을 할일이 없어나 할수없다고 잘못아는 경우가 많은데
    자바스크립트가 어떤 로직을 제어하는 것을 싱글스레드로 비동기적으로 제어할 뿐, 병렬처리는 필요할 때가 있다.
    (ex. node에서 db에 쿼리를 병렬적으로 보내거나 redis같은 nosql에 여러 키를 기준으로 여러값을 가져오거나 이미지를 처리하는 등...)
    그런 병렬처리를 node가 모두 하는 것이 아닌 네트워크나 기타 I/O로 작업을 보낸 후 대기하며 시점을 다루는 일만 node만 하는 것이기 때문에 자바스크립트에서도 병렬처리가 가능하다.
  */
</script>

<script>
  // (동시성으로 동작하는) 병렬처리 사례

  const delay500 = a => new Promise(resolve => {
    console.log('hi-', a)
    setTimeout(() => resolve(a), 500)
  })

  go([1, 2, 3, 4, 5], // 5키를 가지고 redis로부터 값을 가져오는 경우라 생각해보자..
    L.map(a => delay500(a * a)),
    L.filter(a => a % 2),
    reduce(add), // 현재는 지연평가되다가 reduce에 의해 1부터 하나씩 꺼내지면서 평가되어 총 500ms * 5 시간이 걸리니 이를 병렬적으로 처리해보자
    log)
</script>

<script>
  // 함수형 프로그래밍에서는 상태, 외부상황에 의존하여 작동하는게 아닌 코드들(함수들)이 독립적으로 동작하기 떄문에 병렬처리 코드 작성이 쉽다
  const C = {}
  C.reduce = curry((f, acc, iter) => iter ?
    reduce(f, acc, [...iter]) : /// [...iter]로 처리하여 비동기 상황을 제어하지않고 모두 실행시키고 reduce에서 순회하며 값을 꺼낸다. (하나씩 꺼내던 기존 reduce와 다르다)
    reduce(f, [...acc]))

  // go([1, 2, 3, 4, 5], // 5키를 가지고 redis로부터 값을 가져오는 경우라 생각해보자..
  //   L.map(a => delay500(a * a)),
  //   L.filter(a => a % 2),
  //   C.reduce(add), // 현재는 지연평가되다가 reduce에 의해 1부터 하나씩 꺼내지면서 평가되어 총 500ms * 5 시간이 걸리니 이를 병렬적으로 처리해보자
  //   log)
</script>

<script>
  // go([1, 2, 3, 4, 5], // 5키를 가지고 redis로부터 값을 가져오는 경우라 생각해보자..
  //   L.map(a => delay500(a * a)),
  //   L.filter(a => delay500(a % 2)),
  //   L.map(a => delay500(a * a)),
  //   C.reduce(add), // 현재는 지연평가되다가 reduce에 의해 1부터 하나씩 꺼내지면서 평가되어 총 500ms * 5 시간이 걸리니 이를 병렬적으로 처리해보자
  //   log)
  // 위 코드가 계산은 잘되는데도 nop Error 발생, 
  // 자바스크립트 언어의 특징인데, Promise.reject('error')를 그냥 실행해보면 Uncaught 에러 출력 
  // 콜스택에 Promise.reject으로 평가되는 코드가 있으면 위와 같이 출력되도록 되어있기 떄문에..(나중에 이 에러를 처리해도 이미 에러가 찍힌상황)
  /*
    const p = Promise.reject('ho') 
    p.catch(a => console.log('해경'))) // 이렇게 원하는 시점에 catch하려고하도 윗줄에서 출력된 Uncaught 에러는 어쩔 수 없다
  */
</script>

<script>
    /*
      iter2 = iter.map(a => a.catch(function() {})) 와 같이 catch를 한 Promise를 전달하면 추후에 catch가 불가능하기떄문에
      iter2.forEach(a => a.catch(function() {})) 와 같이 catch를 하지 않은 Promise를 전달하되 임시적인 catch를 달아두기만 한 것이다.
      const a = Promise.reject('j1')
      a = a.catch(a => a)를 해버리면 이후에 a.catch(a => console.log(a))를 못한다..

      const a = Promise.reject('j1')
      a.catch(a => a)를 하면 에러는 나오지않고 추후에 원하는 시점에 a.catch(a => console.log(a)) 가능
    */

    function noop_() {} // 아무 처리 안하는 함수를 이렇게 선언하여 많이 사용한다.
    const catchNoop_ = arr => 
      (arr.forEach(a => a instanceof Promise ? a.catch(noop_) : a), arr) // catchNoop 처리 후 arr 그대로 return
    
    C.reduce = curry((f, acc, iter) => {
      const iter2 = catchNoop_(iter ? [...iter] : [...acc])
      return iter ?
        reduce(f, acc, iter2) : // [...iter] 이부분에서 Promise.reject이 평가될 수 있는 상황이기 때문에
        reduce(f, iter2)
    })

    go([1, 2, 3, 4, 5],
    L.map(a => delay500(a * a)),
    L.filter(a => delay500(a % 2)),
    L.map(a => delay500(a * a)),
    C.reduce(add),
    log) // 이제 위의 go...에서 뜨던 에러가 안뜬다.
</script>

<script>
  C.take = curry((l, iter) => take(l, catchNoop_([...iter])))
  // 받을 값을 모두 즉시 (병렬적으로) 평가하되 noop처리 하도록

  go([1, 2, 3, 4, 5],
    L.map(a => delay500(a * a)),
    L.filter(a => delay500(a % 2)),
    L.map(a => delay500(a * a)),
    C.take(2),
    log)
</script>